courses/advanced-ada/chapters/generics.rst

```diff
@@ -408,3 +392,1511 @@ two:
+       begin
+          if Header /= "" then
+             Put_Line (Header);
+             New_Line;
+          end if;
+
+          for E of Container.V loop
+             for F of Fields loop
+                Put (Image (E, F) & " ");
+             end loop;
+             New_Line;
+          end loop;
+
+          New_Line;
+       end Display;
+
+    end Data;
+
+Finally, we implement the :ada:`Test_Data_Container` procedure, which
+makes use of the data container:
+
+.. code:: ada run_button
+
+    with Ada.Strings.Unbounded;   use Ada.Strings.Unbounded;
+    with Ada.Calendar.Formatting;
+
+    with Data;          use Data;
+    with Data_Elements; use Data_Elements;
+
+    procedure Test_Data_Container is
+
+       package App_Data_Container is
+
+          --
+          --  Data container for all operations.
+          --
+          C : Data_Container;
+
+          --
+          --  Display procedures are specific for the
+          --  data container.
+          --
+
+          procedure Display_First_Name_Age is new
+            Display (Container => C,
+                     Fields    => (1 => First_Name_F,
+                                   2 => Age_F),
+                     Header    => "FIRST_NAME AGE");
+
+          procedure Display_Name_Birthday is new
+            Display (Container => C,
+                     Fields    => (1 => First_Name_F,
+                                   2 => Last_Name_F,
+                                   3 => Birthday_F),
+                     Header    => "NAME BIRTHDAY");
+       end App_Data_Container;
+
+       use App_Data_Container;
+
+       --
+       --  Data container initialization
+       --
+
+       procedure Init_Container is
+          function To_US (S : String) return Unbounded_String renames
+            To_Unbounded_String;
+       begin
+          Insert (C, (First_Name => To_US ("John"),
+                      Last_Name  => To_US ("Smith"),
+                      Birthday   => Ada.Calendar.Formatting.Time_Of
+                        (Year        => 1951,
+                         Month       => 5,
+                         Day         => 1)));
+
+          Insert (C, (First_Name => To_US ("Alice"),
+                      Last_Name  => To_US ("Williams"),
+                      Birthday   => Ada.Calendar.Formatting.Time_Of
+                        (Year        => 1968,
+                         Month       => 10,
+                         Day         => 12)));
+       end Init_Container;
+
+    begin
+       Init_Container;
+
+       Display_First_Name_Age;
+       Display_Name_Birthday;
+
+    end Test_Data_Container;
+
+In this example, we declare the data container :ada:`C` and bind it to
+two instantiations of the :ada:`Display` procedure:
+
+- :ada:`Display_First_Name_Age`, which displays the first name and age of
+  each person from the database;
+
+- :ada:`Display_Name_Birthday`, which displays the full name and birthday
+  of each person.
+
+Formal access types
```

Comment:
what you describe below is not formal access types (see http://www.ada-auth.org/standards/rm12_w_tc1/html/RM-12-5-4.html) but formal object parameters of access (to subprogram) type. I'm not sure this section adds anything useful.